src/ApiCore/ArrayTrait.php

```diff
@@ -0,0 +1,113 @@
+<?php
+/*
+ * Copyright 2017, Google Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+namespace Google\ApiCore;
+
+/**
+ * Provides basic array helper methods.
+ */
+trait ArrayTrait
+{
+    /**
+     * Pluck a value out of an array.
+     *
+     * @param string $key
+     * @param array $arr
+     * @param bool $isRequired
+     * @return string|null
+     * @throws \InvalidArgumentException
+     */
+    private function pluck($key, array &$arr, $isRequired = true)
+    {
+        if (!array_key_exists($key, $arr)) {
+            if ($isRequired) {
+                throw new \InvalidArgumentException(
```

Comment:
Should this be `ValidationException`?

In general, should all exceptions we throw be ones that we have defined, either `ValidationException` or `ApiException`? This seems like a question of what is standard practice in PHP, so I am happy to stick with `InvalidArgumentException` if that is the standard.

Reply:
I've seen projects that like to extend `InvalidArgumentException` under their own namespace, and others which just use the original SPL exception. I would say the standard exists at the codebase level.

For a little more detail, this trait comes from [google-cloud-php](https://github.com/GoogleCloudPlatform/google-cloud-php/blob/master/src/Core/ArrayTrait.php). I ported it here as I found a few use cases where it helped to keep the code a bit more succinct. The plan is to remove the guts of the code in google-cloud-php and just extend it from here. Given that, it would be nice to retain `InvalidArgumentException` for google-cloud-php so there are no breaking changes - but we could just convert the `ValidationException` over in google-cloud-php as a way retain BC.

Personally, I like the granularity of using the standard SPL exceptions instead of just a blanket `ValidationException` for anything that goes wrong - but would find it acceptable either way. Which do you prefer?

Reply:
I am happy to stick with `InvalidArgumentException`